File: python/tests/test_relay.py
"""Tests for cross-channel relay middleware."""

from __future__ import annotations

import pytest
import pytest_asyncio

from unified_channel.relay import RelayMiddleware, RelayRule
from unified_channel.types import (
    ContentType,
    Identity,
    MessageContent,
    OutboundMessage,
    UnifiedMessage,
)


async def _noop_handler(msg: UnifiedMessage):
    return None


async def _ok_handler(msg: UnifiedMessage):
    return "ok"


def _make_msg(
    channel: str = "telegram",
    text: str = "hello",
    sender_id: str = "user1",
    chat_id: str = "chat1",
    msg_id: str = "m1",
) -> UnifiedMessage:
    return UnifiedMessage(
        id=msg_id,
        channel=channel,
        sender=Identity(id=sender_id, username=sender_id, display_name=f"User {sender_id}"),
        content=MessageContent(type=ContentType.TEXT, text=text),
        chat_id=chat_id,
        raw={},
    )


class FakeAdapter:
    """Mock adapter that records sent messages."""

    def __init__(self, channel_id: str = "slack"):
        self.channel_id = channel_id
        self.sent: list[OutboundMessage] = []

    async def send(self, msg: OutboundMessage) -> str | None:
        self.sent.append(msg)
        return None


class FakeManager:
    """Mock manager with adapters."""

    def __init__(self):
        self.adapters: dict[str, FakeAdapter] = {}

    def add_adapter(self, adapter: FakeAdapter) -> None:
        self.adapters[adapter.channel_id] = adapter

    def get_adapter(self, channel_id: str):
        return self.adapters.get(channel_id)


@pytest.mark.asyncio
async def test_basic_relay():
    """Messages from telegram should relay to slack."""
    relay = RelayMiddleware()
    relay.add_rule("telegram", "slack", target_chat_id="general")

    slack = FakeAdapter("slack")
    mgr = FakeManager()
    mgr.add_adapter(slack)
    relay.set_manager(mgr)

    msg = _make_msg(channel="telegram", text="test message")
    result = await relay.process(msg, _ok_handler)
    assert result == "ok"
    assert len(slack.sent) == 1
    assert "test message" in slack.sent[0].text
    assert slack.sent[0].chat_id == "general"


@pytest.mark.asyncio
async def test_relay_includes_sender():
    """Relayed message should include sender info by default."""
    relay = RelayMiddleware()
    relay.add_rule("telegram", "slack", target_chat_id="ch1")

    slack = FakeAdapter("slack")
    mgr = FakeManager()
    mgr.add_adapter(slack)
    relay.set_manager(mgr)

    msg = _make_msg(channel="telegram", text="hello world", sender_id="alice")
    await relay.process(msg, _noop_handler)

    assert "[telegram/User alice]" in slack.sent[0].text
    assert "hello world" in slack.sent[0].text


@pytest.mark.asyncio
async def test_relay_without_sender():
    """When include_sender=False, no sender prefix."""
    relay = RelayMiddleware()
    relay.add_rule("telegram", "slack", target_chat_id="ch1", include_sender=False)

    slack = FakeAdapter("slack")
    mgr = FakeManager()
    mgr.add_adapter(slack)
    relay.set_manager(mgr)

    msg = _make_msg(text="raw text")
    await relay.process(msg, _noop_handler)

    assert slack.sent[0].text == "raw text"


@pytest.mark.asyncio
async def test_relay_filter():
    """Only messages passing filter should be relayed."""
    relay = RelayMiddleware()
    relay.add_rule(
        "telegram",
        "email",
        target_chat_id="admin@example.com",
        filter_fn=lambda m: "urgent" in (m.content.text or "").lower(),
    )

    email_adapter = FakeAdapter("email")
    mgr = FakeManager()
    mgr.add_adapter(email_adapter)
    relay.set_manager(mgr)

    # Normal message — not relayed
    msg1 = _make_msg(text="just chatting")
    await relay.process(msg1, _noop_handler)
    assert len(email_adapter.sent) == 0

    # Urgent message — relayed
    msg2 = _make_msg(text="URGENT: server down", msg_id="m2")
    await relay.process(msg2, _noop_handler)
    assert len(email_adapter.sent) == 1
    assert "server down" in email_adapter.sent[0].text


@pytest.mark.asyncio
async def test_relay_transform():
    """Custom transform function should modify relayed text."""
    relay = RelayMiddleware()
    relay.add_rule(
        "telegram",
        "slack",
        target_chat_id="alerts",
        transform=lambda m: f"ALERT: {m.content.text}",
        include_sender=False,
    )

    slack = FakeAdapter("slack")
    mgr = FakeManager()
    mgr.add_adapter(slack)
    relay.set_manager(mgr)

    msg = _make_msg(text="CPU at 95%")
    await relay.process(msg, _noop_handler)

    assert slack.sent[0].text == "ALERT: CPU at 95%"


@pytest.mark.asyncio
async def test_wildcard_source():
    """Wildcard '*' source matches all channels."""
    relay = RelayMiddleware()
    relay.add_rule("*", "slack", target_chat_id="firehose")

    slack = FakeAdapter("slack")
    mgr = FakeManager()
    mgr.add_adapter(slack)
    relay.set_manager(mgr)

    for ch in ["telegram", "discord", "email", "twilio_sms"]:
        msg = _make_msg(channel=ch, text=f"from {ch}", msg_id=ch)
        await relay.process(msg, _noop_handler)

    assert len(slack.sent) == 4
    assert "telegram" in slack.sent[0].text
    assert "discord" in slack.sent[1].text


@pytest.mark.asyncio
async def test_broadcast():
    """Broadcast sends to multiple targets."""
    relay = RelayMiddleware()
    relay.add_broadcast(
        "telegram",
        {"slack": "general", "email": "team@co.com", "discord": "12345"},
    )

    slack = FakeAdapter("slack")
    email_a = FakeAdapter("email")
    discord_a = FakeAdapter("discord")
    mgr = FakeManager()
    mgr.add_adapter(slack)
    mgr.add_adapter(email_a)
    mgr.add_adapter(discord_a)
    relay.set_manager(mgr)

    msg = _make_msg(text="broadcast test")
    await relay.process(msg, _noop_handler)

    assert len(slack.sent) == 1
    assert len(email_a.sent) == 1
    assert len(discord_a.sent) == 1
    assert slack.sent[0].chat_id == "general"
    assert email_a.sent[0].chat_id == "team@co.com"


@pytest.mark.asyncio
async def test_bidirectional():
    """Bidirectional rule creates both directions."""
    relay = RelayMiddleware()
    relay.add_rule("telegram", "slack", target_chat_id="ch1", bidirectional=True)

    slack = FakeAdapter("slack")
    telegram = FakeAdapter("telegram")
    mgr = FakeManager()
    mgr.add_adapter(slack)
    mgr.add_adapter(telegram)
    relay.set_manager(mgr)

    # Telegram → Slack
    msg1 = _make_msg(channel="telegram", text="from tg")
    await relay.process(msg1, _noop_handler)
    assert len(slack.sent) == 1

    # Slack → Telegram
    msg2 = _make_msg(channel="slack", text="from slack", msg_id="m2")
    await relay.process(msg2, _noop_handler)
    assert len(telegram.sent) == 1


@pytest.mark.asyncio
async def test_no_relay_for_unmatched_source():
    """Messages from unmatched channels are not relayed."""
    relay = RelayMiddleware()
    relay.add_rule("telegram", "slack", target_chat_id="ch1")

    slack = FakeAdapter("slack")
    mgr = FakeManager()
    mgr.add_adapter(slack)
    relay.set_manager(mgr)

    msg = _make_msg(channel="discord", text="should not relay")
    await relay.process(msg, _noop_handler)

    assert len(slack.sent) == 0


@pytest.mark.asyncio
async def test_relay_metadata():
    """Relayed messages include metadata about the source."""
    relay = RelayMiddleware()
    relay.add_rule("telegram", "slack", target_chat_id="ch1")

    slack = FakeAdapter("slack")
    mgr = FakeManager()
    mgr.add_adapter(slack)
    relay.set_manager(mgr)

    msg = _make_msg(channel="telegram", msg_id="orig123")
    await relay.process(msg, _noop_handler)

    meta = slack.sent[0].metadata
    assert meta["relayed_from"] == "telegram"
    assert meta["original_id"] == "orig123"


@pytest.mark.asyncio
async def test_relay_missing_adapter():
    """Relay to missing adapter should not crash."""
    relay = RelayMiddleware()
    relay.add_rule("telegram", "nonexistent", target_chat_id="ch1")

    mgr = FakeManager()  # no adapters
    relay.set_manager(mgr)

    msg = _make_msg()
    # Should not raise
    result = await relay.process(msg, _ok_handler)
    assert result == "ok"


@pytest.mark.asyncio
async def test_relay_adapter_error():
    """Relay send failure should not break the handler chain."""
    relay = RelayMiddleware()
    relay.add_rule("telegram", "slack", target_chat_id="ch1")

    class BrokenAdapter:
        channel_id = "slack"
        async def send(self, msg):
            raise ConnectionError("network down")

    mgr = FakeManager()
    mgr.adapters["slack"] = BrokenAdapter()
    relay.set_manager(mgr)

    msg = _make_msg()
    result = await relay.process(msg, _ok_handler)
    assert result == "ok"


@pytest.mark.asyncio
async def test_chaining_api():
    """add_rule returns self for fluent chaining."""
    relay = (
        RelayMiddleware()
        .add_rule("telegram", "slack", target_chat_id="ch1")
        .add_rule("slack", "email", target_chat_id="a@b.com")
    )
    assert len(relay._rules) == 2

File: rust/tests/test_relay.rs
//! Tests for RelayMiddleware: rules, filters, transforms, broadcast, bidirectional.

use std::collections::HashMap;
use std::sync::Arc;

use unified_channel::middleware::*;
use unified_channel::relay::*;
use unified_channel::types::*;

fn make_msg(channel: &str, sender: &str, text: &str) -> UnifiedMessage {
    let mut msg = UnifiedMessage::new(
        channel,
        Identity::new(sender).with_display_name(sender.to_string()),
        MessageContent::text(text),
    );
    msg.chat_id = Some("chat1".to_string());
    msg
}

#[tokio::test]
async fn test_relay_basic_rule() {
    let mut relay = RelayMiddleware::new();
    relay.add_rule("telegram", "slack", "general", None, None, true, false);
    assert_eq!(relay.rule_count(), 1);
}

#[tokio::test]
async fn test_relay_bidirectional_adds_two_rules() {
    let mut relay = RelayMiddleware::new();
    relay.add_rule("telegram", "slack", "general", None, None, true, true);
    assert_eq!(relay.rule_count(), 2);
}

#[tokio::test]
async fn test_relay_broadcast_adds_multiple_rules() {
    let mut relay = RelayMiddleware::new();
    let mut targets = HashMap::new();
    targets.insert("slack".to_string(), "general".to_string());
    targets.insert("discord".to_string(), "lobby".to_string());
    targets.insert("email".to_string(), "team@co.com".to_string());
    relay.add_broadcast("telegram", targets, None, None);
    assert_eq!(relay.rule_count(), 3);
}

#[tokio::test]
async fn test_relay_passes_through_result() {
    // RelayMiddleware should pass through the next handler's result
    let mut relay = RelayMiddleware::new();
    relay.add_rule("telegram", "slack", "general", None, None, true, false);

    let handler: Handler = handler_fn(|_| async { HandlerResult::Text("original".into()) });
    let msg = make_msg("telegram", "alice", "hello");
    let result = relay.process(msg, &handler).await;
    assert!(matches!(result, HandlerResult::Text(s) if s == "original"));
}

#[tokio::test]
async fn test_relay_with_filter() {
    let filter: FilterFn = Arc::new(|msg: &UnifiedMessage| {
        msg.content.text.contains("urgent")
    });

    let mut relay = RelayMiddleware::new();
    relay.add_rule("telegram", "slack", "alerts", Some(filter), None, true, false);
    assert_eq!(relay.rule_count(), 1);

    // The relay processes but the filter is checked internally
    let handler: Handler = handler_fn(|_| async { HandlerResult::None });

    // Non-matching message
    let msg = make_msg("telegram", "alice", "hello");
    let result = relay.process(msg, &handler).await;
    assert!(result.is_none());

    // Matching message
    let msg = make_msg("telegram", "alice", "urgent: server down");
    let result = relay.process(msg, &handler).await;
    assert!(result.is_none()); // relay doesn't change the handler result
}

#[tokio::test]
async fn test_relay_with_transform() {
    let transform: TransformFn = Arc::new(|msg: &UnifiedMessage| {
        format!("SUMMARY: {}", msg.content.text.chars().take(20).collect::<String>())
    });

    let mut relay = RelayMiddleware::new();
    relay.add_rule("telegram", "slack", "digests", None, Some(transform), false, false);
    assert_eq!(relay.rule_count(), 1);
}

#[tokio::test]
async fn test_relay_wildcard_source() {
    let mut relay = RelayMiddleware::new();
    relay.add_rule("*", "telegram", "log_chat", None, None, true, false);
    assert_eq!(relay.rule_count(), 1);

    // Wildcard should match any channel
    let handler: Handler = handler_fn(|_| async { HandlerResult::None });

    let msg = make_msg("discord", "bob", "hello");
    let result = relay.process(msg, &handler).await;
    assert!(result.is_none());

    let msg = make_msg("slack", "carol", "world");
    let result = relay.process(msg, &handler).await;
    assert!(result.is_none());
}

#[tokio::test]
async fn test_relay_no_matching_rules() {
    let mut relay = RelayMiddleware::new();
    relay.add_rule("slack", "telegram", "123", None, None, true, false);

    let handler: Handler = handler_fn(|_| async { HandlerResult::Text("ok".into()) });

    // Message from discord, no rules match
    let msg = make_msg("discord", "user", "hello");
    let result = relay.process(msg, &handler).await;
    assert!(matches!(result, HandlerResult::Text(s) if s == "ok"));
}

#[tokio::test]
async fn test_relay_default() {
    let relay = RelayMiddleware::default();
    assert_eq!(relay.rule_count(), 0);
}

#[tokio::test]
async fn test_relay_chained_add_rules() {
    let mut relay = RelayMiddleware::new();
    relay
        .add_rule("telegram", "slack", "general", None, None, true, false)
        .add_rule("slack", "email", "team@co.com", None, None, true, false)
        .add_rule("*", "telegram", "logs", None, None, false, false);
    assert_eq!(relay.rule_count(), 3);
}
